gravitee-am-repository/gravitee-am-repository-mongodb/src/main/java/io/gravitee/am/repository/mongodb/management/MongoRateLimitRepository.java

```diff
@@ -0,0 +1,162 @@
+import io.reactivex.Completable;
+import io.reactivex.Maybe;
+import io.reactivex.Observable;
+import io.reactivex.Single;
+import org.bson.Document;
+import org.bson.conversions.Bson;
+import org.springframework.stereotype.Component;
+
+import javax.annotation.PostConstruct;
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.mongodb.client.model.Filters.and;
+import static com.mongodb.client.model.Filters.eq;
+
+/**
+ * @author Ashraful Hasan (ashraful.hasan at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+@Component
+public class MongoRateLimitRepository extends AbstractManagementMongoRepository implements RateLimitRepository {
+    private static final String RATE_LIMIT = "rate_limit";
+    private static final String FIELD_FACTOR_ID = "factorId";
+
+    private MongoCollection<RateLimitMongo> rateLimitCollection;
+
+    @PostConstruct
+    public void init() {
+        rateLimitCollection = mongoOperations.getCollection(RATE_LIMIT, RateLimitMongo.class);
+        super.init(rateLimitCollection);
+        super.createIndex(rateLimitCollection, new Document(FIELD_USER_ID, 1)
+                .append(FIELD_CLIENT, 1)
+                .append(FIELD_FACTOR_ID,1));
+    }
+
+    @Override
+    public Maybe<RateLimit> findById(String id) {
+        return Observable.fromPublisher(rateLimitCollection.find(eq(FIELD_ID, id)).first()).firstElement().map(this::convert);
+    }
+
+    @Override
+    public Maybe<RateLimit> findByCriteria(RateLimitCriteria criteria) {
+        return Observable.fromPublisher(rateLimitCollection.find(query(criteria)).first()).firstElement().map(this::convert);
+    }
+
+    @Override
+    public Single<RateLimit> create(RateLimit item) {
+        RateLimitMongo rateLimitMongo = convert(item);
+        return Single.fromPublisher(rateLimitCollection.insertOne(rateLimitMongo)).flatMap(success -> {
+            item.setId(rateLimitMongo.getId());
+            return Single.just(item);
+        });
+    }
+
+    @Override
+    public Single<RateLimit> update(RateLimit item) {
+        RateLimitMongo rateLimitMongo = convert(item);
+        return Single.fromPublisher(rateLimitCollection.replaceOne(eq(FIELD_ID, rateLimitMongo.getId()), rateLimitMongo)).flatMap(success -> Single.just(item));
+    }
+
+    @Override
+    public Completable delete(String id) {
+        return Completable.fromPublisher(rateLimitCollection.deleteOne(eq(FIELD_ID, id)));
+    }
+
+    @Override
+    public Completable delete(RateLimitCriteria criteria) {
+        return Completable.fromPublisher(rateLimitCollection.deleteOne(query(criteria)));
+    }
+
+    @Override
+    public Completable deleteByUser(String userId) {
+        return Completable.fromPublisher(rateLimitCollection.deleteMany(eq(FIELD_USER_ID, userId)));
+    }
+
+    @Override
+    public Completable deleteByDomain(String domainId) {
+        return Completable.fromPublisher(rateLimitCollection.deleteMany(eq(FIELD_REFERENCE_ID, domainId)));
```

Comment:
Maybe we can add here the ReferenceType.DOMAIN as second criteria.

Reply:
done, added as suggested